File: Minimal string.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"

int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	string str;
	cin >> str;
	vector<char> v(str.size() + 1);
	v[str.size()] = 'z';
	for (int i = str.size()-1; i >= 0; i--)
		v[i] = min(v[i+1], str[i]);
	int i = 0;
	stack<char> st;
	while (!st.empty() || i < str.size()) {
		if (!st.empty() && st.top() <= v[i]) {
			cout << st.top();
			st.pop();
		} else {
			st.push(str[i++]);
		}
	}
	cout << endl;
}

File: Minimal string.py
s = input()
c = ['z'] * (len(s) + 1)
for i in range(len(s)-1, -1, -1):
	c[i] = min(c[i+1], s[i])
st = list()
i = 0
while len(st) > 0 or i < len(s):
	if len(st) > 0 and st[len(st)-1] <= c[i]:
		print(st[len(st)-1], end = "")
		st.pop()
	else:
		st.append(s[i])
		i += 1
print()
